src/parser/flow_lexer.ml

```diff
@@ -724,65 +767,86 @@ let token (env: Lex_env.t) lexbuf : result =
       is_tail
     ))
 
-  | binnumber, (letter | '2'..'9'), Star alphanumeric ->
+  | binbigint ->
+    Token (env, T_BIGINT { kind = BINARY; raw = lexeme lexbuf })
+  | (binnumber | binbigint), (letter | '2'..'9'), Star alphanumeric ->
     (* Numbers cannot be immediately followed by words *)
     recover env lexbuf ~f:(fun env lexbuf -> match%sedlex lexbuf with
     | binnumber ->
       Token (env, T_NUMBER { kind = BINARY; raw = lexeme lexbuf })
+    | binbigint ->
+      Token (env, T_BIGINT { kind = BINARY; raw = lexeme lexbuf })
     | _ -> failwith "unreachable"
     )
-
   | binnumber ->
     Token (env, T_NUMBER { kind = BINARY; raw = lexeme lexbuf })
 
-  | octnumber, (letter | '8'..'9'), Star alphanumeric ->
+  | octbigint ->
+    Token (env, T_BIGINT { kind = OCTAL; raw = lexeme lexbuf })
+  | (octnumber | octbigint), (letter | '8'..'9'), Star alphanumeric ->
     (* Numbers cannot be immediately followed by words *)
     recover env lexbuf ~f:(fun env lexbuf -> match%sedlex lexbuf with
     | octnumber -> Token (env, T_NUMBER { kind = OCTAL; raw = lexeme lexbuf })
+    | octbigint ->
+      Token (env, T_BIGINT { kind = OCTAL; raw = lexeme lexbuf })
     | _ -> failwith "unreachable"
     )
-
   | octnumber ->
     Token (env, T_NUMBER { kind = OCTAL; raw = lexeme lexbuf })
 
-  | legacyoctnumber, (letter | '8'..'9'), Star alphanumeric ->
+  | legacyoctbigint ->
+    Token (env, T_BIGINT { kind = LEGACY_OCTAL; raw = lexeme lexbuf })
+  | (legacyoctnumber | legacyoctbigint), (letter | '8'..'9'), Star alphanumeric ->
     (* Numbers cannot be immediately followed by words *)
     recover env lexbuf ~f:(fun env lexbuf -> match%sedlex lexbuf with
     | legacyoctnumber -> Token (env, T_NUMBER { kind = LEGACY_OCTAL; raw = lexeme lexbuf })
+    | legacyoctbigint -> Token (env, T_BIGINT { kind = LEGACY_OCTAL; raw = lexeme lexbuf })
     | _ -> failwith "unreachable"
     )
-
   | legacyoctnumber ->
     Token (env, T_NUMBER { kind = LEGACY_OCTAL; raw = lexeme lexbuf })
 
-  | hexnumber, non_hex_letter, Star alphanumeric ->
+  | hexbigint ->
+    Token (env, T_BIGINT { kind = NORMAL; raw = lexeme lexbuf })
+  | (hexnumber | hexbigint), non_hex_letter, Star alphanumeric ->
     (* Numbers cannot be immediately followed by words *)
     recover env lexbuf ~f:(fun env lexbuf -> match%sedlex lexbuf with
     | hexnumber -> Token (env, T_NUMBER { kind = NORMAL; raw = lexeme lexbuf })
+    | hexbigint -> Token (env, T_BIGINT { kind = NORMAL; raw = lexeme lexbuf })
     | _ -> failwith "unreachable"
     )
-
   | hexnumber ->
     Token (env, T_NUMBER { kind = NORMAL; raw = lexeme lexbuf })
 
-  | scinumber, word ->
+  | scibigint ->
+    let loc = loc_of_lexbuf env lexbuf in
+    let env = lex_error env loc Parse_error.InvalidBigInt in
+    Token (env, T_BIGINT { kind = NORMAL; raw = lexeme lexbuf })
+  | (scinumber | scibigint), word ->
```

Comment:
It's actually ok for this to be the same case. However, it DOES need to come before both the `| scibigint` and `| scinumber` cases.